kore-parser: Make --no-print-definition and --no-print-pattern default and make --module and --pattern mandatory to go together

MirceaS opened the pull request:
Fixes #2264 
Fixes #2254 

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

ttuegel reviewed (changes requested):
Please add unit tests, particularly to check that the combination `--no-print-definition --print-definition` works. You will probably have to move `enableDisableFlag` into the library.

I know that command-line parsing is my favorite example of low-value tests, but I'm starting to change my mind; it's obviously not simple to get command-line parsing right and the tests are easy to write.

MirceaS commented:
I updated the type of the parsed options so that it better reflects what we want but since this would break existing code, I aslo added a transformed from the new type to the old.
I believe that code using this type should be refactored so that it uses the new type I introduced, at which point we can get rid of the transformer (`morph`) but if that proves to be too difficult/time consuming we may safely leave it in.

MirceaS commented on kore/app/parser/Main.hs:
```diff
@@ -164,21 +103,23 @@ main = handleTop $ do
                         $ toVerifiedDefinition indexedModules
                 else putDoc (debug parsedDefinition)
 
-            let KoreParserOptions { patternFileName } = koreParserOptions
-            unless (null patternFileName) $ do
-                parsedPattern <- mainPatternParse patternFileName
-                when willVerify $ do
-                    let KoreParserOptions { mainModuleName } = koreParserOptions
-                    indexedModule <-
-                        lookupMainModule
-                            (ModuleName mainModuleName)
-                            indexedModules
-                        & lift
-                    _ <- mainPatternVerify indexedModule parsedPattern
-                    return ()
-                let KoreParserOptions { willPrintPattern } = koreParserOptions
-                when willPrintPattern $
-                    lift $ putDoc (debug parsedPattern)
```
This is just a code refactoring so that it uses the new `KoreParserOptions` type that I introduced

ttuegel commented on kore/app/parser/Main.hs:
```diff
@@ -164,21 +103,23 @@ main = handleTop $ do
-                when willPrintPattern $
-                    lift $ putDoc (debug parsedPattern)
+            let KoreParserOptions { patternOpt } = koreParserOptions
+            case patternOpt of
+                Nothing -> pure ()
+                Just PatternOptions {patternFileName, mainModuleName} -> do
+                    parsedPattern <- mainPatternParse patternFileName
+                    when willVerify $ do
+                        indexedModule <-
+                            lookupMainModule
+                                (ModuleName mainModuleName)
+                                indexedModules
+                            & lift
+                        _ <- mainPatternVerify indexedModule parsedPattern
+                        return ()
```
```suggestion
            Foldable.for_ patternOpt $ \patternOptions -> do
                let PatternOptions { patternFileName } = patternOptions
                parsedPattern <- mainPatternParse patternFileName
                when willVerify $ do
                    let PatternOptions { mainModuleName } = patternOptions
                    indexedModule <-
                        lookupMainModule
                            (ModuleName mainModuleName)
                            indexedModules
                        & lift
                    _ <- mainPatternVerify indexedModule parsedPattern
                    return ()
```

ttuegel commented on kore/src/Kore/OptionsParser.hs:
```diff
@@ -0,0 +1,112 @@
+
+-}
+
+module Kore.OptionsParser
```
Please rename this module `Kore.Options` and make it re-export `Options.Applicative`.

ttuegel commented on kore/src/Kore/OptionsParser.hs:
```diff
@@ -0,0 +1,112 @@
+        (  long ( "[no-]" ++ name )
+        <> help ( "Enable/disable " ++ helpSuffix ) )
+    <|> pure defaultVal
+
```
```suggestion
```

ttuegel commented on kore/src/Kore/OptionsParser.hs:
```diff
@@ -0,0 +1,112 @@
+    }
+
+-- | Command Line Argument Parser
+commandLineParser :: Parser KoreParserOptions
+commandLineParser =
```
```suggestion
parseKoreParserOptions :: Parser KoreParserOptions
parseKoreParserOptions =
```

ttuegel commented on kore/src/Kore/OptionsParser.hs:
```diff
@@ -0,0 +1,112 @@
+        (  metavar "FILE"
+        <> help "Kore source file to parse [and verify]" )
+    <*> (   (Just <$>
+                (PatternOptions
+                <$> strOption
+                    (  metavar "PATTERN_FILE"
+                    <> long "pattern"
+                    <> help "Kore pattern source file to parse [and verify]. Needs --module.")
+                <*> strOption
+                    (  metavar "MODULE"
+                    <> long "module"
+                    <> help "The name of the main module in the Kore definition")
+                )
```
Please extract this as `parsePatternOptions`.

ttuegel commented on kore/src/Kore/OptionsParser.hs:
```diff
@@ -0,0 +1,112 @@
+    <$> argument str
+        (  metavar "FILE"
+        <> help "Kore source file to parse [and verify]" )
+    <*> (   (Just <$>
+                (PatternOptions
+                <$> strOption
+                    (  metavar "PATTERN_FILE"
+                    <> long "pattern"
+                    <> help "Kore pattern source file to parse [and verify]. Needs --module.")
+                <*> strOption
+                    (  metavar "MODULE"
+                    <> long "module"
+                    <> help "The name of the main module in the Kore definition")
+                )
+            )
+            <|> flag Nothing Nothing
+                (  long "definition-only"
+                <> help "No Kore pattern may be checked against the definition")
+        )
```
```suggestion
    <*> optional parsePatternOptions
```

ttuegel commented on kore/test/Test/Kore/OptionsParser.hs:
```diff
@@ -0,0 +1,86 @@
+test_flags =
+    [ testGroup "print-definition"
+        [ testCase "default is False" $ do
+            let flag_value = willPrintDefinition $ runParser commandLineParser
```
Please follow the style guide:

```suggestion
            let flagValue = willPrintDefinition $ runParser commandLineParser
```
